### PR TITLE
pytest(macOS): Run pytest tests on macOS

### PR DESCRIPTION
**Pull request by echoix**
This is a PR from work started (continued from a previous attempts months before) at the very end of the community meeting. This enables pytest tests to run on macOS, an easier step towards running more tests using pytest instead of gunittest.

The reason it isn’t exactly straightforward to simply run the same commands on every platform is because our test suite (at least for the pytest side) uses some constructs that work correctly only using the "fork" start method with multiprocessing. Since it is possible to be dangerous, the default start method on POSIX will change from "fork" to "spawn" in Python 3.14. On macOS, the default start method changed to "spawn" in Python 3.8 without much notice since "fork" was broken with OS changes at the time. On Windows, "fork" is simply not available.

What are the differences between the start methods, at least for "spawn"? Spawn is slower than fork, but is safer. Spawn needs to pickle the data and spawn a completely new interpreter with its own memory. Fork will use a fork call and the new interpreter will have the memory of the parent interpreter at the moment the fork was made. It is possible to corrupt some memory, and it’s not recommended to use when also using threads, where deadlocks can occur. I highly suspect that we were encountering these deadlocks when some tests using pytest-xdist ran some tests that used multiple threads/procs, making other tests timeout. That’s why we had some pytest tests split in two steps for a couple months. From what I recall, I haven’t restarted any CI runs because of timeouts since the last time I added some pytest markers. Actually solving these deadlocks will be needed to run these tests on macOS and Windows, and this PR is a start.
There’s also a third startup method, "fork server", but isn’t as relevant, as we still need spawn to work for Windows. macOS has a faster feedback cycle (especially if we cache the mamba env, it takes 18 seconds, I removed it for this PR).

The sentence that says that spawn uses pickle is very important. Pickling has limitations, and we are hitting some. For one, pickle can’t find local functions (functions defined inside a function). Lambdas can have the same problem (it’s easier to say it doesn’t work at all, even if not completely true). Sometimes a little more reorganization must be made if side effects happen when executing a file again, like correctly defining some parts to run with ˋif __name__ == "__main__"ˋ. 

So, about these changes:
- Changes conditional imports (for optional dependencies) to use pytest.importorskip(). This pattern was already discussed and agreed upon in June, with @wenzeslaus among others. A PR just for that wasn’t filed yet, but the work for macOS was based on it. Let me know if you’d like a more pedantic process and would like to block this PR until a second one with only these changes is merged and this PR updated again. It works perfectly fine and only one importorskip is needed for the file, without adding decorators to each test.
- Added expected failures for tests that won’t work if the start method is "spawn". I defined it once per file and reused the decorator, so changing the string is easier. I couldn’t place it in a conftest.py file, but it could be placed in an importable file and be imported. To have less import problems during testing, I didn’t go further with this.
- Added one example of how functools.partial can be used to replace a local function with a function that is defined outside, but has some fixed arguments, but some come from the context of the test function, and ends up acting as a function handle (a function assigned in a variable) with less or no arguments, and that can be used with multiprocessing and with the spawn start method. It is pretty much the accepted solution for this problem, instead of completely changing the code.
- Used the if ${{ !cancelled() }} pattern used in our other CI workflows, to align with them (as some parts were used as is, so it’s good to have them aligned). They were useful when I made lots of runs, but I didn’t remove it when I did the last rebase that is used in this PR (I have a more advanced branch of this work, that is already a reduction of the commits of the last 2 months or so). I couldn’t place remove it if you strictly need a separate PR for it, but I don’t think it’s required.

What’s to come soon (what I already done but was cut off for this PR, so in further PRs once this one is merged):
- Adding pytest code coverage:
  - for the single worker test step, no real coverage is made yet, so useless. 
  - For combined code coverage of the two pytest steps (it works)
  - For uploading to codecov, the codecov step needs to use gpg (from gnupg), but that node-based action calls it directly through with /bin/sh, and thus can’t find the one installed through mamba. Same thing for coverage.py tool, it doesn’t seem to like a conda env. (I know from other work that I can bypass that by creating the json file myself before using coverage.py that is available when using the conda env, perhaps removing the .coverage data file so it doesn’t think it needs to convert it). But for the gpg, it still doesn’t really work because of the changes @nilason made for macos-14, of completely removing homebrew. I was able to install homebrew back at the end and install gnupg that is already available on the runners usually. There’s also a flaw in that move+delete step: homebrew used /opt/local when it was using intel cpus, and they decided to use /opt/homebrew for Apple Silicon so a same binary that runs with Rosetta 2 can also exist when the system has a Apple silicon version. That explains why moving both folders to a same /opt/off folder has errors that a subdirectory isn’t empty (2 times), as they can in fact have similar contents. Either way I don’t understand exactly why it was needed to hide homebrew completely, but it pretty much breaks homebrew for the rest of the CI run after that step.
- Adding more functools.partial versions of the xfail-ed tests. I made some for the pygrass/modules/grid tests, and almost worked, but required some other changes in other functions throughout Python code, that use local functions even if not needed. (In handle_errors, in grass.script.core, in grass.script.raster, and more). I still had one remaining failure that needs some changes, but I don’t understand exactly why the « slope » map isn’t found (is it an error or the consequence of another error of a non-pickled function?)
- In the functools.partial based versions of the tests, there’s a way to not have to have a separate inner function defined in the outer scope with only the changed arguments, so only using one for all the variations of the tests, by defining the outer function to accept ˋ*args, **kwargs` and defining everything in functools.partial. It doesn’t test exactly the same thing (not inside a function), but for the module/function called, it’s the same. Also, if for example GridModule also defined `__call__ˋ instead of needing to use .run(), it might be possible to only use partial without an outer function at all. It’s to follow.
- (Not done yet): run pytest on Windows. The last time I tried, (started months before the community meeting), I was blocked on not able to ~~(angry bad word)~~ find pytest only in GH CI even if it was installed through OSGeo4W and able to see that the files are there. Maybe something that interfered with the path resolutions and hides it, anyways I moved on after a while (and long feedback cycles on CI)

Finishing thoughts: I was surprised how so little was left to do to make it work, I was so close to having it work when I stopped working on it, as I was blocked. But it only took a small idea and was able to make it work (from my phone only) the same evening, before cleaning up the multi-months draft version with lots and lots of rebases.

### CODE REVIEW

**Comment by echoix** [comment]
Oh boy, I wrote so much!

File: python/grass/script/tests/grass_script_setup_test.py [hunk 1]
```diff
@@ -62,6 +70,32 @@ def test_init_finish_global_functions_with_env(tmp_path):
...
+        init_finish_global_functions_capture_strerr0_partial, tmp_path
+    )
+    session_file, runtime_present = run_in_subprocess(init_finish)
+    assert session_file, "Expected file name from the subprocess"
+    assert runtime_present, "Runtime (GISBASE) should be present"
+    assert not os.path.exists(session_file), "Session file not deleted"
+
+
+@xfail_mp_spawn
 def test_init_finish_global_functions_capture_strerr0(tmp_path):
```
Comment (echoix) [inline python/grass/script/tests/grass_script_setup_test.py:99]: @wenzeslaus what does strerr0 mean? It got changed to it in one of your PRs. If it was intended to be stderr, I would change the name of the new test I made.

File: .github/workflows/macos_dependencies.txt [hunk 2]
```diff
@@ -1,7 +1,7 @@
 blas
 cairo
-clangxx_osx-arm64
```
Comment (nilason) [inline .github/workflows/macos_dependencies.txt:3]: You may revert this unrelated change. (It was sorted alphabetically).

File: .github/workflows/macos_dependencies.txt [hunk 3]
```diff
@@ -11,7 +11,6 @@ gettext
 ghostscript
 giflib
 git
-libjpeg-turbo
```
Comment (nilason) [inline .github/workflows/macos_dependencies.txt:14]: Before I saw there was a duplicate of this dependant, I thought this was not necessary... Good catch!

**Review by nilason — CHANGES_REQUESTED** [review]
With the exception of the unrelated re-sorting, the mac part of this looks ok. I'll leave the python part for someone more up to-date knowledge on pytest.

File: .github/workflows/macos_dependencies.txt (hunk 2 above)
Comment (echoix) [reply .github/workflows/macos_dependencies.txt:3]: I also used the F9 sort in VSCode + delete duplicate lines when doing the dozen or so rebases.. to simplify it I just kept that order in the meantime.
Was yours sorted by hand or automatically?

**Comment by nilason** [comment]
You're right the clean-up of homebrew is not correct. I did notice this, but haven't prioritised addressing that. Yet.
It is not advisable to keep homebrew mixed with the conda build we have. Might lead to unexpected problems. There is a conda `gnupg` package if needed.

**Comment by echoix** [comment]
And @nilason I'll need your help at the next step on how to have gnupg from homebrew available at the end, considering that you explicitly removed it at the start of the job

File: .github/workflows/macos_dependencies.txt (hunk 2 above)
Comment (nilason) [reply .github/workflows/macos_dependencies.txt:3]: Automatically. I suppose there is a different sorting algorithm for non-alphabetical/digit characters.

**Comment by echoix** [comment]
> You're right the clean-up of homebrew is not correct. I did notice this, but haven't prioritised addressing that. Yet.
> 
> It is not advisable to keep homebrew mixed with the conda build we have. Might lead to unexpected problems. There is a conda `gnupg` package if needed.

But the tool is in an action, in node, and doesn't activate the environment, so doesn't find it

File: .github/workflows/macos.yml [hunk 4]
```diff
@@ -55,12 +55,31 @@ jobs:
...
-        if: always()
+        if: ${{ !cancelled() }}
         shell: bash -l {0}
         run: source ./.github/workflows/print_versions.sh
-      - name: Run tests
+
+      - name: Run pytest with multiple workers in parallel
+        shell: bash -el {0}
+        run: |
+          export PYTHONPATH=`grass --config python_path`:$PYTHONPATH
```
Comment (nilason) [inline .github/workflows/macos.yml:65]:
```suggestion
          export PYTHONPATH=$(grass --config python_path):$PYTHONPATH
```

File: .github/workflows/macos.yml [hunk 5]
```diff
@@ -55,12 +55,31 @@ jobs:
...
+          export LD_LIBRARY_PATH=$(grass --config path)/lib:$LD_LIBRARY_PATH
+          pytest --verbose --color=yes --durations=0 --durations-min=0.5 \
+            --numprocesses auto \
+            -ra . \
+            -m 'not needs_solo_run'
+      - name: Run pytest with a single worker (for tests marked with needs_solo_run)
         shell: bash -el {0}
-        run: >
+        run: |
+          export PYTHONPATH=`grass --config python_path`:$PYTHONPATH
```
Comment (nilason) [inline .github/workflows/macos.yml:74]:
```suggestion
          export PYTHONPATH=$(grass --config python_path):$PYTHONPATH
```

**Review by nilason — CHANGES_REQUESTED** [review]
Sorry, found a couple of minor things.

If needed for the tests, please consider add `ipython` and `pyyaml` conda packages to the dependencies file.

File: .github/workflows/macos.yml (hunk 4 above)
Comment (echoix) [reply .github/workflows/macos.yml:65]: Should these be (in other PRs) ported to our other workflows? It is aligned with our other CI steps, so that they aren't diverging from the main pytest workflow.

Comment (nilason) [reply .github/workflows/macos.yml:65]: Yes, eventually. `$(...)` is to be preferred to using backticks. (I believe shellcheck also recommends this.).

**Comment by echoix** [comment]
> Sorry, found a couple of minor things.
> 
> 
> 
> If needed for the tests, please consider add `ipython` and `pyyaml` conda packages to the dependencies file.
> 
> 

I can try to put it, but I think it's a good idea to limit the scope and show that the importorskip works, and add more afterwards in subsequent PRs. It was deliberately a minimal PR that was conservative, in order to be merged quickly (it seems it didn't work).

If ever the actual tests don't work (I wouldn't be surprised as it never ran), I would not include it in this PR to not block it.

File: .github/workflows/macos.yml (hunk 4 above)
Comment (nilason) [reply .github/workflows/macos.yml:65]: I think it's better to only change here for now.

**Comment by nilason** [comment]
A more general question: what is the relation between pytest and unittest? Do they overlap, can pytest replace the other?

File: .github/workflows/macos.yml (hunk 4 above)
Comment (echoix) [reply .github/workflows/macos.yml:65]: On macOS, the default shell isn’t bash. On GitHub CI, we set the shell for the step to a bash (that’s why it would work). But if shell wasn’t set, is this syntax also supported?

**Comment by echoix** [comment]
> A more general question: what is the relation between pytest and unittest? Do they overlap, can pytest replace the other?

In almost any other project, yes, but since we are so different and our grass Python code isn't set up as a Python package and importable etc, it can't (yet). It's the yearlong goal I'm working on, and I need some PRs to be merged at some points for it. Coverage work is also similar, in order to prove that both approaches test the same thing, and see what isn't when switching.

I want both to be equivalent before continuing to reduce gunittest. And to be sure changes in the packaging/importing are correct, testing needs to be robust. It was longly discussed with Vaclav.
In short, that's a small step, it's not a complete endeavour by itself.

**Comment by echoix** [comment]
The goal is to not invest that much new time in gunittest. But gunittest is also lagging with respect to current python's unittest. And gunittest wasn't always subclassed as much as it could.

**Comment by echoix** [comment]
@nilason I can add ipython and pyyaml, but tests using ipywidgets are not ready yet, they fail with strange errors:

https://github.com/echoix/grass/actions/runs/10376884034/job/28730312672?pr=191


The first two are fine https://github.com/echoix/grass/actions/runs/10376740425/job/28729357776

File: .github/workflows/macos.yml (hunk 4 above)
Comment (nilason) [reply .github/workflows/macos.yml]: sh, bash, zsh are so related to each other that in this respect this is nothing that makes a difference. Backticks works for all, but $() is recommended (see https://www.shellcheck.net/wiki/SC2006).

**Review by nilason — APPROVED** [review]
The mac part looks good to me. Thanks!

**Comment by echoix** [comment]
Even though it is "ready to merge", I'm leaving the opportunity to comment on these:
- The text of the message that is repeated
- What does "strerr0" in the test title mean. Is it intended to be "stderr0" (it doesn't make any more sense to me because of the 0 at the end).

File: python/grass/script/tests/grass_script_setup_test.py (hunk 1 above)
Comment (wenzeslaus) [reply python/grass/script/tests/grass_script_setup_test.py:99]: While this does not strike me as an typo, I don't see nor remember a reason why this should be there, so let's treat that as a typo. If you feel like changing the other ones while doing this overall improvement, please go ahead.

Comment (echoix) [reply python/grass/script/tests/grass_script_setup_test.py:99]: Should my new test simply not have strerr0 then?

Comment (echoix) [reply python/grass/script/tests/grass_script_setup_test.py:99]: I'll do the renaming on its own, so it's clear that the test are paired, until we have a test only for the different ways of calling the functions, add a duplicate of the tests with our preferred calling method (for cross platform tests), wait a bit, and remove the older tests after some time.

**Review by wenzeslaus — COMMENTED** [review]
Just to add to the review by @nilason, the Python code looks good. I don't have a strong opinion about the message.

File: python/grass/script/tests/grass_script_setup_test.py (hunk 1 above)
Comment (wenzeslaus) [reply python/grass/script/tests/grass_script_setup_test.py:99]: Sounds good. Maybe the names are just different variations of the same test, i.e., better name is needed.
